src/base-widget.ts

```diff
@@ -3,27 +3,44 @@ import { isPlatformBrowser } from '@angular/common';
 import { bem, noop } from './utils';
 import { NgAisInstantSearch } from './instantsearch/instantsearch';
 import { NgAisIndex } from './index-widget/index-widget';
-import { Widget } from 'instantsearch.js/es/types';
-export { Widget };
-
-// TODO: use Connector type from InstantSearch. Not yet possible now,
-// since non-ts connectors can't have generics like Connector has,
-// as well as sometimes being not accurate enough / missing keys.
-export type Connector = (
-  renderFn: (state: object, isFirstRendering: boolean) => void,
-  unmountFn: () => void
-) => (widgetOptions?: object) => Widget;
-
-export abstract class BaseWidget implements OnInit, OnDestroy {
+import {
+  Widget,
+  WidgetDescription,
+  Connector,
+} from 'instantsearch.js/es/types';
+
+export { Widget, Connector };
+
+export type ExtractConnectorParams<
+  T extends Connector<unknown & WidgetDescription, unknown>
+> = T extends Connector<infer TWidgetDescription, infer TConnectorParams>
+  ? TConnectorParams
+  : never;
+
+export type ExtractRenderer<
+  T extends Connector<unknown & WidgetDescription, unknown>
+> = Parameters<T>[0];
+
+export type ExtractUnmounter<
+  T extends Connector<unknown & WidgetDescription, unknown>
+> = Parameters<T>[1];
+
+export type ExtractRenderState<
+  T extends Connector<unknown & WidgetDescription, unknown>
+> = Parameters<ExtractRenderer<T>>[0];
+
+export abstract class BaseWidget<
+  TConnector extends Connector<unknown & WidgetDescription, unknown>
+> implements OnInit, OnDestroy {
   @Input() public autoHideContainer?: boolean;
 
   public widget?: Widget;
```

Comment:
```suggestion
  public widget?: Widget | IndexWidget;
```

this wasn't needed?

Reply:
I started off like that to migrate the Index widget but then I had issues with the rest because it doesn't have a connector, a state and a renderFn. I tried forging a connectIndex that worked, but it looked a little hacky and started looking for another solution.

I was wondering if it would be simpler solution to just not use `BaseWidget`, and inline the parts that sense for Index. Pros are
- There are other widgets that don't:  `Panel`, `Highlight`
- Index is not using much of the BaseWidget machinery (no bem classes, no state, no updateState, no `autoHideContainer`

Here how IndexWidget would look like if it didn't use BaseWidget. 


```diff
-  export class NgAisIndex extends BaseWidget
+  export class NgAisIndex implements OnInit, OnDestroy {
   @Input() public indexName: IndexWidgetParams['indexName'];
   @Input() public indexId?: IndexWidgetParams['indexId'];
 
   public widget?: IndexWidget;
 
   constructor(
     // public API does not include SkipSelf, but the index widget should accept parents, avoiding itself.
     @SkipSelf()
     @Inject(forwardRef(() => NgAisIndex))
     @Optional()
     public parentIndex: NgAisIndex,
     @Inject(forwardRef(() => NgAisInstantSearch))
     public instantSearchInstance: NgAisInstantSearch
   ) {
-   super('Index');
   }
 
+  get parent() {
+    if (this.parentIndex) {
+      return this.parentIndex;
+    }
+    return this.instantSearchInstance;
+  }
 
   public addWidgets(widgets: Widget[]) {
     this.widget.addWidgets(widgets);
   }
 
   public removeWidgets(widgets: Widget[]) {
     this.widget.removeWidgets(widgets);
   }
 
+  public ngOnInit() {
+    this.widget = indexWidget({
+      indexName: this.indexName,
+      indexId: this.indexId,
+    });
+    this.parent.addWidgets([this.widget]);
+  }
+  public ngOnDestroy() {
+    if (isPlatformBrowser(this.instantSearchInstance.platformId)) {
+      this.parent.removeWidgets([this.widget]);
+    }
+  }
 }
```
Isn't this simpler than stretching the types in BaseWidget?

Reply:
that's a good idea, in Vue InstantSearch we did "extend base widget" as it's mostly the same, but yeah let's go with not making an exception to baseWidget for index.